### PR TITLE
Cache symbol completion result

### PR DESCRIPTION
**Pull request by tompng**
Make symbol completion faster

- Cache `Symbol.all_symbols.sort` and reuse it.
- Restrict the number of symbol candidates to reduce completion/rendering cost in Reline.
- Use bsearch to perform O(logN) search from cached sorted symbol list.

```
irb -f
irb(main):001> s='a';symbols = 100000.times.map{s=s.succ; [s.to_sym, ('a'+s).to_sym]};
irb(main):002> :aaabq
               :aaabk 
               :aaabl 
               :aaabm 
               :aaabn 
               :aaabo 
               :aaabp 
               :aaabq▄
               :azyd █
               :azye ▀
               :azyf  
               :azyg  
               :azyh  
               :azyi  
               :azyj  
               :azyk  
```
Completion candidates of symbol contains first 50 + last 50 symbols matched to the completion target.

### CODE REVIEW

File: lib/irb/completion.rb [hunk 1]
```diff
@@ -116,18 +139,27 @@ def completion_candidates(preposing, target, _postposing, bind:)
...
 
+    def type_completion_candidates(preposing, target, bind)
       result = ReplTypeCompletor.analyze(preposing + target, binding: bind, filename: @context.irb_path)
+      return [] unless result
 
-      return commands unless result
-
-      commands | result.completion_candidates.map { target + _1 }
+      analyze_result = result.instance_variable_get(:@analyze_result)
+      if analyze_result.is_a?(Array) && analyze_result[0] == :symbol && analyze_result[1].is_a?(String)
```
Comment (tompng) [inline lib/irb/completion.rb:157]: I think it is relatively safe because it checks the value strictly, but it's a bit hacky...

**Review by st0012 — COMMENTED** [review]
I feel sorry to say this because I was the one who brought up this idea 😅 But are we sure that the performance improvement is worth the complexity? Especially the binary search part.

**Comment by tompng** [comment]
I think there is room for discussion whether to keep symbol completion or to drop it.

If we're going to keep it, we need to do this bsearch because symbol filtering (main bottleneck in IRB side) is few times slower than `Symbol.all_symbols`.

File: lib/irb/completion.rb [hunk 2]
```diff
@@ -96,6 +96,29 @@ def command_candidates(target)
...
+    end
+
+    def symbol_candidates(prefix, first: 50, last: 50)
+      limit = first + last
+      symbol_names = @sorted_symbol_names ||= Symbol.all_symbols.filter_map do
+        _1.inspect[1..]
+      rescue EncodingError
+        # ignore
+      end.sort
+      start_index = symbol_names.bsearch_index { |sym| sym.to_s >= prefix }
```
[the remark this answers is not in the document]
Comment (tompng) [reply lib/irb/completion.rb]: Sorry `symbol_names` was already a string. I deleted `to_s` and add missing `return [] unless start_index`.
We need to search from an array of string `symbols.map(&:inspect)` to exclude symbols like `:"a b"`.

**Comment by tompng** [comment]
In the following situation, `Symbol.all_symbol.sort` takes about a second. I think it is not acceptable.

Symbol.all_symbols.size is 100000 with average symbol length=15 (About the number in Rails app)
Running in docker (host and image platform mismatch)

**Pull request closed without merging**